ArduCopter/control_safe_rtl.cpp

```diff
@@ -0,0 +1,168 @@
+#include "Copter.h"
+
+/*
+ * Init and run calls for Safe_RTL flight mode
+ *
+ * This code uses the SafeRTL path that is already in memory, and feeds it into WPNav, one point at a time.
+ * Once the copter is close to home, it will run a standard land controller.
+ */
+
+bool Copter::safe_rtl_init(bool ignore_checks)
+{
+    if ((position_ok() || ignore_checks) && safe_rtl_path.is_active()) {
+        // initialise waypoint and spline controller
+        wp_nav->wp_and_spline_init();
+        wp_nav->init_loiter_target();
+
+        Vector3f current_pos;
+        ahrs.get_relative_position_NED_origin(current_pos);
+        // if we're within 4m of the landing point, just skip right to pre-land-positioning
+        // otherwise, begin a normal safertl procedure from the start
+        if (HYPOT(current_pos, safe_rtl_path.get(0)) <= 4.0f) {
+            current_pos[2] -= 2.0f; // go to the point two meters above the landing spot, then start landing.
+            safe_rtl_state = SafeRTL_PreLandPosition;
+        } else {
+            safe_rtl_state = SafeRTL_WaitForCleanup;
+        }
+
+        // set current position as the target point.
+        DataFlash.Log_Write_SRTL(DataFlash_Class::SRTL_POINT_GOTO, current_pos);
+        wp_nav->set_wp_destination_NED_origin(current_pos);
+
+        // initialise yaw to obey user parameter
+        set_auto_yaw_mode(get_default_auto_yaw_mode(true));
+
+        // tell library to stop accepting new breadcrumbs
+        safe_rtl_path.accepting_new_points = false;
+        return true;
+    }else{
+        return false;
+    }
+}
+
+void Copter::safe_rtl_run()
+{
+    switch(safe_rtl_state){
+        case SafeRTL_WaitForCleanup:
+            safe_rtl_wait_cleanup_run();
+            break;
+        case SafeRTL_PathFollow:
+            safe_rtl_path_follow_run();
+            break;
+        case SafeRTL_PreLandPosition:
+            safe_rtl_pre_land_position_run();
+            break;
+        case SafeRTL_Descend:
+            rtl_descent_run(); // Re-using the descend method from normal rtl mode.
+            break;
+        case SafeRTL_Land:
+            rtl_land_run(); // Re-using the land method from normal rtl mode.
+            break;
+    }
+}
+
+void Copter::safe_rtl_wait_cleanup_run()
+{
+    wp_nav->update_wpnav();
+    pos_control->update_z_controller();
+    attitude_control->input_euler_angle_roll_pitch_yaw(wp_nav->get_roll(), wp_nav->get_pitch(), get_auto_heading(),true, get_smoothing_gain());
+
+    if (safe_rtl_path.cleanup_ready()){
+        safe_rtl_path.thorough_cleanup();
+        safe_rtl_state = SafeRTL_PathFollow;
+    }
+}
+
+void Copter::safe_rtl_path_follow_run()
+{
+    // if we are close to current target point, switch the next point to be our target.
+    if(wp_nav->reached_wp_destination()){
+        Vector3f next_point;
+        bool last_point = safe_rtl_path.pop_point(next_point);
```

Comment:
Perhaps have a `point_count` instead of having `pop_point` return whether this is the last point; bool return here would ordinarily indicate whether you successfully popped a point,